### PR TITLE
Adds contracts for byte_add, byte_sub and byte_offset

### PR DESCRIPTION
**Pull request by stogaru**
- Also adds proofs verifying unit, int and composite types.

> Please add a description of your PR.
> If this is a solution to an open challenge, please explain your solution.
>
> Don't forget to check our book to ensure your solution satisfy the overall
> requirements as well as the challenge success criteria.
>

Resolves #ISSUE-NUMBER

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/ptr/mut_ptr.rs [hunk 1]
```diff
@@ -2302,3 +2378,157 @@ impl<T: ?Sized> PartialOrd for *mut T {
...
+                }
+            }
+        };
+    }
+
+    gen_mut_byte_arith_unit_harness!(byte_add, check_mut_byte_add_unit);
+    gen_mut_byte_arith_unit_harness!(byte_sub, check_mut_byte_sub_unit);
+    gen_mut_byte_arith_unit_harness!(byte_offset, check_mut_byte_offset_unit);
+
+    const ARRAY_LEN: usize = 40;
```
Comment (xsxszab) [inline library/core/src/ptr/mut_ptr.rs:2399]: We need to document magic numbers, even if it looks trivial

File: library/core/src/ptr/mut_ptr.rs [hunk 2]
```diff
@@ -94,7 +99,10 @@ impl<T: ?Sized> *mut T {
     /// // This dereference is UB. The pointer only has provenance for `x` but points to `y`.
     /// println!("{:?}", unsafe { &*bad });
     #[unstable(feature = "set_ptr_value", issue = "75091")]
-    #[cfg_attr(bootstrap, rustc_const_stable(feature = "ptr_metadata_const", since = "1.83.0"))]
+    #[cfg_attr(
+        bootstrap,
+        rustc_const_stable(feature = "ptr_metadata_const", since = "1.83.0")
+    )]
```
Comment (xsxszab) [inline library/core/src/ptr/mut_ptr.rs]: The rustfmt formatting is unnecessary

File: library/core/src/ptr/mut_ptr.rs [hunk 3]
```diff
@@ -277,7 +285,11 @@ impl<T: ?Sized> *mut T {
     pub const unsafe fn as_ref<'a>(self) -> Option<&'a T> {
         // SAFETY: the caller must guarantee that `self` is valid for a
         // reference if it isn't null.
-        if self.is_null() { None } else { unsafe { Some(&*self) } }
+        if self.is_null() {
```
Comment (xsxszab) [inline library/core/src/ptr/mut_ptr.rs]: The rustfmt formatting is unnecessary

File: library/core/src/ptr/mut_ptr.rs [hunk 4]
```diff
@@ -352,7 +364,11 @@ impl<T: ?Sized> *mut T {
     {
         // SAFETY: the caller must guarantee that `self` meets all the
         // requirements for a reference.
-        if self.is_null() { None } else { Some(unsafe { &*(self as *const MaybeUninit<T>) }) }
+        if self.is_null() {
```
Comment (xsxszab) [inline library/core/src/ptr/mut_ptr.rs]: The rustfmt formatting is unnecessary

File: library/core/src/ptr/mut_ptr.rs [hunk 5]
```diff
@@ -537,7 +569,9 @@ impl<T: ?Sized> *mut T {
     #[stable(feature = "pointer_byte_offsets", since = "1.75.0")]
     #[rustc_const_stable(feature = "const_pointer_byte_offsets", since = "1.75.0")]
     pub const fn wrapping_byte_offset(self, count: isize) -> Self {
-        self.cast::<u8>().wrapping_offset(count).with_metadata_of(self)
+        self.cast::<u8>()
```
Comment (xsxszab) [inline library/core/src/ptr/mut_ptr.rs]: The rustfmt formatting is unnecessary

File: library/core/src/ptr/mut_ptr.rs [hunk 6]
```diff
@@ -578,7 +612,9 @@ impl<T: ?Sized> *mut T {
     #[must_use = "returns a new pointer rather than modifying its argument"]
     #[inline(always)]
     pub fn mask(self, mask: usize) -> *mut T {
-        intrinsics::ptr_mask(self.cast::<()>(), mask).cast_mut().with_metadata_of(self)
+        intrinsics::ptr_mask(self.cast::<()>(), mask)
```
Comment (xsxszab) [inline library/core/src/ptr/mut_ptr.rs]: The rustfmt formatting is unnecessary

File: library/core/src/ptr/mut_ptr.rs [hunk 7]
```diff
@@ -628,7 +664,11 @@ impl<T: ?Sized> *mut T {
     pub const unsafe fn as_mut<'a>(self) -> Option<&'a mut T> {
         // SAFETY: the caller must guarantee that `self` is be valid for
         // a mutable reference if it isn't null.
-        if self.is_null() { None } else { unsafe { Some(&mut *self) } }
+        if self.is_null() {
```
Comment (xsxszab) [inline library/core/src/ptr/mut_ptr.rs]: The rustfmt formatting is unnecessary

File: library/core/src/ptr/mut_ptr.rs [hunk 8]
```diff
@@ -689,7 +729,11 @@ impl<T: ?Sized> *mut T {
     {
         // SAFETY: the caller must guarantee that `self` meets all the
         // requirements for a reference.
-        if self.is_null() { None } else { Some(unsafe { &mut *(self as *mut MaybeUninit<T>) }) }
+        if self.is_null() {
```
Comment (xsxszab) [inline library/core/src/ptr/mut_ptr.rs]: The rustfmt formatting is unnecessary

**Comment by stogaru** [comment]
https://github.com/model-checking/verify-rust-std/pull/169

**Pull request closed without merging**